### ISSUE TITLE
The key controller.current.ctor was not bound to any value.

### ISSUE THREAD

**Issue by NiklasA** [strongloop/loopback-next#1144]
## Description / Steps to reproduce / Feature proposal:
Hi everyone, 
I added @loopback/authentication to my LoopBack 4 application (like descriped in [README.md](https://github.com/strongloop/loopback-next/blob/master/packages/authentication/README.md) and I always get the error: "The key controller.current.ctor was not bound to any value.". Does anyone know what I have to do, any hint, where I have to look? Thanks in advance!

## Current Behavior:
Every time, when I call my TestController: `curl -u username:password http://localhost:3000/whoami`. I get the following error: Error: The key controller.current.ctor was not bound to any value.

**Comment by raymondfeng** [comment]
@NiklasA It would be helpful if you can either describe steps to reproduce the problem or having a sandbox repo to contain the test app.

**Comment by NiklasA** [comment]
Hi @raymondfeng,

I found my issue. The problem was that I integrated `@loopback/authentication` and `example-log-extension` both and that I merged both snippets wrong in `sequence.ts`. I had to move `await this.authenticateRequest(req);` into try / catch block.

**Before:**
```javascript
  async handle(req: ParsedRequest, res: ServerResponse) {
    // we define these variable outside of try / catch so that they can be accessed by logger.
    let args: any = [];
    let result: any;

    // this is the important line added to the default sequence implementation.
    await this.authenticateRequest(req);

    // optionally start timing the sequence using the timer function available via LogFn.
    const start = this.logger.startTimer();

    try {
      const route = this.findRoute(req);
``` 

**After:**
```javascript
  async handle(req: ParsedRequest, res: ServerResponse) {
    // we define these variable outside of try / catch so that they can be accessed by logger.
    let args: any = [];
    let result: any;

    // optionally start timing the sequence using the timer function available via LogFn.
    const start = this.logger.startTimer();

    try {
      const route = this.findRoute(req);

      // this is the important line added to the default sequence implementation.
      await this.authenticateRequest(req);
```

The second mistake was, that I injected `@inject(RestBindings.Http.CONTEXT) public ctx: Context` in constructor of MySequence. Do you know, why it is not possible to inject Context, when I use @loopback/authentication?

Thank you in advance.
Niklas

**Comment by raymondfeng** [comment]
Please note we use a hierarchy of contexts, for example, a request context is created per http request and it uses the app context as the parent. Binding `controller.current.ctor` is only available for the request ctx. The injection of context at constructor level will get the app context.

**Comment by bajtos** [comment]
@raymondfeng @dhmlau @NiklasA Hello, what's the status of this issue? What actions are needed before we can closed this as done?

**Comment by NiklasA** [comment]
Hi @bajtos, it can be closed, it made a mistake, described in [comment above](https://github.com/strongloop/loopback-next/issues/1144#issuecomment-374766232).

**Comment by bajtos** [comment]
Good catch, @NiklasA. I was aware that we may break the applications assuming that 404 is thrown by `findRoute` when we change our router to throw the error from `invoke`, but wasn't sure. However, I think this isn't fixable at our side, since the authentication tutorial is correctly showing that all sequence actions should be wrapped in the try/catch block, not the `findRoute` action only.

Perhaps we should improve the documentation on routing and `findRoute` action to make it clear that when static assets are registered, then 404 is thrown by `invoke`?

Most importantly, I think we need to fix our reference `authenticateRequest` function to correctly support `StaticAssetsRoute` and any other custom route that apps or extensions my register via `app.route()` API.

Is my assessment correct?

@dhmlau I am proposing to open new issue(s) for each of the logical change/fix we need to make, and then close this issue again. Having a fine-grained user stories is easier for planning and keeping track of our progress.

**Comment by basitsattar** [comment]
@jormar anwered solved my issue too.

**Comment by rpvelloso** [comment]
same problem here, @jormar answer solved it.

**Comment by joaozitopolo** [comment]
it's VERY important the @jormar information about static routes. The code at https://github.com/strongloop/loopback-next/tree/master/packages/authentication worked before the recent update of loopback 4, that now the root is mapped to /public.

**Comment by jannyHou** [comment]
FYI this should be fixed in https://github.com/strongloop/loopback-next/pull/2218

Released as @loopback/authentication@1.0.9
Let me know if updating to that version still doesn't solve the bug.

**Comment by jannyHou** [comment]
I am closing this issue as solved. Feel free to reopen it if you still run into the same problem
Please see solution details in the comment above ^.
Thanks.
